tests/handlers/test_sso.py

```diff
@@ -0,0 +1,127 @@
+from tests.test_utils import SMALL_PNG
+from tests.test_utils import FakeResponse, simple_async_mock
+from twisted.web.http_headers import Headers
+from typing import BinaryIO, Optional, Callable, Union, Mapping, Tuple, Dict, List
+from synapse.logging.context import make_deferred_yieldable
+from twisted.web.client import readBody
+
+RawHeaders = Union[Mapping[str, "RawHeaderValue"], Mapping[bytes, "RawHeaderValue"]]
+
+
+class TestSSOHandler(unittest.HomeserverTestCase):
+    def make_homeserver(self, reactor: MemoryReactor, clock: Clock) -> HomeServer:
+        self.http_client = Mock(spec=["request", "get_file"])
+        self.http_client.request.side_effect = mock_request
+        self.http_client.get_file.side_effect = mock_get_file
+        self.http_client.user_agent = b"Synapse Test"
+        hs = self.setup_test_homeserver(proxied_http_client=self.http_client)
+        return hs
+
+    async def test_set_avatar(self):
+        """Tests successfully setting the avatar of a newly created user"""
+        handler = self.hs.get_sso_handler()
+
+        # TODO: Create a new user to set avatar for
+        reg_handler = self.hs.get_registration_handler()
+        user_id = await reg_handler.register_user(approved=True)
+        # user_id = "@sso-user:test"
+
+        with self.assertLogs() as cm:
+            await handler.set_avatar(user_id, "http://my.server/me.png")
+        self.assertEqual(
+            cm.output, ["INFO:synapse.handlers.sso:successfully saved the user avatar"]
+        )
+
+        # TODO: ensure avatar returned via user's profile is SMALL_PNG
+        profile_handler = self.hs.get_profile_handler()
+        profile = await profile_handler.get_profile(user_id)
+        # profile["avatar_url"]
+
+    @unittest.override_config({"max_avatar_size": 99})
+    async def test_set_avatar_too_big_image(self):
+        """Tests saving of avatar failed when image size is too big"""
+        handler = self.hs.get_sso_handler()
+
+        # any random user works since image check is supposed to fail
+        user_id = "@sso-user:test"
+
+        with self.assertLogs() as cm:
+            await handler.set_avatar(user_id, "http://my.server/big.png")
+        self.assertEqual(
+            cm.output, ["WARNING:synapse.handlers.sso:failed to save the user avatar"]
+        )
+
+    @unittest.override_config({"allowed_avatar_mimetypes": ["image/jpeg"]})
+    async def test_set_avatar_incorrect_mime_type(self):
+        """Tests saving of avatar failed when not allowed mimetype of image was used"""
+        handler = self.hs.get_sso_handler()
+
+        # any random user works since image check is supposed to fail
+        user_id = "@sso-user:test"
+
+        with self.assertLogs() as cm:
+            await handler.set_avatar(user_id, "http://my.server/me.png")
+        self.assertEqual(
+            cm.output, ["WARNING:synapse.handlers.sso:failed to save the user avatar"]
+        )
+
+
+async def mock_request(method: str, url: str):
```

Comment:
Can you add a return type annotation here please? I'd guess this is `-> Awaitable[FakeResponse]` but am not sure.

Can you also include a `disallow_untyped_defs True` entry for this file in mypy.ini?

Reply:
Done!